### PR TITLE
GCN Multi-Class Node Classification - s4642798

### PR DESCRIPTION
**Pull request by WILLJB02**
**Name:** William Barker
**Student Number:** 46427986

This pull request implements a comprehensive solution for the Multi-Class Node Classification of the [Facebook Network Dataset](https://snap.stanford.edu/data/facebook-large-page-page-network.html) using a Graph Convolutional Network (GCN). The PR includes an in-depth README documenting the task, model, dataset, usage and efficacy (found in the `recognition\GCN_s4642798` folder).  For convenience, this is summarized below:

**Task:** Build a Graph Convolutional Network to classify the Facebook pages in the [Facebook Large Page-Page Network](https://snap.stanford.edu/data/facebook-large-page-page-network.html) dataset into one of four categories: politicians, governmental organizations, television shows, and companies.
**Model:** GCN which uses 3 graph convolution layers. ReLU activation function is used after the first two layers, and SoftMax is used after the final layer. ADAM is used as the optimizer, and Cross Entropy is used as the loss function.
**Dataset:** The model uses the partially processed `facebook.npz` file (found [here](https://graphmining.ai/datasets/ptg/facebook.npz)).
**Usage:** Download the `facebook.npz`  dataset and move it to the `recognition\GCN_s4642798` folder. Install dependencies listed within `recognition\GCN_s4642798\README.md`. Run the `train.py` file and then the `predict.py` file from the `recognition\GCN_s4642798` folder.
**Efficacy:** The model achieves 92.82% accuracy on test data that was split from the dataset.
**Files Added:** 
- `README.md`: Report and breakdown of the GCN model and its usage instructions.
- `dataset.py`: Loads the `facebook.npz` file and performs pre-processing.
- `modules.py`: Contains the class for the GCN model.
- `train.py`:  Logic for training the GCN model. 
- `predict.py`: Visualization of the results of the GCN model. 
- `./plots`: Images of the generated plots during training and visualization.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Normal

Readme Overall: Good

- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:

- Multi-layer GCN: Good
- Semi supervised: Good
- Accuracy: Good
- TSNE or UMAP: Good
- Discussion: Good

Code:

- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
Good report. 
You can include an early stop.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
